packages/protocol/contracts/psp/AirnodeSponsorPspAllocator.sol

```diff
@@ -0,0 +1,104 @@
+        address sponsor;
+    }
+
+    IAirnodeRrp public airnodeRrp;
+    AirnodePsp public airnodePsp;
+    mapping(address => mapping(address => uint256))
+        public airnodeToSponsorToNoOfTotalSlots;
+    mapping(address => mapping(address => uint256))
+        public airnodeToSponsorToNoOfUsedSlots;
+    mapping(address => mapping(uint256 => Slot))
+        public airnodeToSlotIndexToSlot;
+
+    constructor(address airnodeRrp_, address airnodePsp_) {
+        airnodeRrp = IAirnodeRrp(airnodeRrp_);
+        airnodePsp = AirnodePsp(airnodePsp_);
+    }
+
+    function setTotalNoOfSlots(
+        address airnode,
+        address sponsor,
+        uint256 noOfTotalSlots
+    ) external {
+        // Check if the caller is an Airnode admin, i.e., this contract will inherit SelfAdminnable
+        // require(msg.sender == ...)
+        airnodeToSponsorToNoOfTotalSlots[airnode][sponsor] = noOfTotalSlots;
+    }
+
+    // Need to come up with a way to prevent people from constantly switching between slots to grief
+    function subscribe(bytes32 subscriptionId, uint256 slotIndex) external {
+        (bytes32 templateId, , , , , ) = airnodePsp.subscriptions(
+            subscriptionId
+        );
+        (address airnode, , ) = airnodeRrp.templates(templateId);
+        require(
+            airnodeToSponsorToNoOfTotalSlots[airnode][msg.sender] >
+                airnodeToSponsorToNoOfUsedSlots[airnode][msg.sender],
+            "Sponsor has no available slots"
+        );
+        address revokedSponsor = airnodeToSlotIndexToSlot[airnode][slotIndex]
+            .sponsor;
+        if (revokedSponsor != address(0)) {
+            require(
+                airnodeToSponsorToNoOfTotalSlots[airnode][revokedSponsor] <
+                    airnodeToSponsorToNoOfUsedSlots[airnode][revokedSponsor],
+                "Slot with index not available"
+            );
+            airnodeToSponsorToNoOfUsedSlots[airnode][revokedSponsor]--;
+        }
+        airnodeToSponsorToNoOfUsedSlots[airnode][msg.sender]++;
+        airnodeToSlotIndexToSlot[airnode][slotIndex] = Slot({
+            subscriptionId: subscriptionId,
+            sponsor: msg.sender
+        });
+    }
+
+    function unsubscribe(address airnode, uint256 slotIndex) external {
+        require(
+            msg.sender == airnodeToSlotIndexToSlot[airnode][slotIndex].sponsor,
+            "Caller does not own slot"
+        );
+        airnodeToSponsorToNoOfUsedSlots[airnode][msg.sender]--;
+        delete airnodeToSlotIndexToSlot[airnode][slotIndex];
+    }
+
+    // At the start of each invocation, the node calls this where `limit` is the number of
+    // slots specified for this allocator in the config file
+    function getActiveSubscriptions(address airnode, uint256 limit)
+        external
+        view
+        returns (bytes32[] memory subscriptionIds,
+        address[] memory sponsors)
+    {
+        // Did a similar thing in the DAO ConvenienceUtils.sol
```

Comment:
I suppose this is pointless, we can just return all the first `limit` items and let the node filter them

Reply:
Any advantage of not doing it in the contract function directly? 

I can imagine an implementation of Allocator where slots will have some automated "time to live" (e.g. block number or timestamp) and this function can filter those "outdated" slots... So I like this function as a part of an "AllocatorInterface"

Reply:
Doing computation in a static call is more expensive than doing the same computation at the node (the blockchain provider will charge you for the compute units and/or the calls will be responded to with more delay)